Pokedex Project

krzysztof-kryczka opened the pull request.

krzysztof-kryczka committed:
installation of necessary lib (react-router-dom, react-hook-form, zod, axios, notistack, clsx, tailwindcss, postcss, autoprefixer) and config for tailwind

krzysztof-kryczka committed:
created custom hook: useFetchPokemons and component view

krzysztof-kryczka committed:
created pagination component and petite styles for pagination

krzysztof-kryczka committed:
refactor pagination component

krzysztof-kryczka committed:
fix: pagination component

krzysztof-kryczka committed:
added additional info from api for pokemon card in Pokedex component and change style

krzysztof-kryczka committed:
added search filtered poke

krzysztof-kryczka committed:
configuration router for app.jsx and main.jsx, added logo, added blank PokemonDetails component

krzysztof-kryczka committed:
modified PokemonDetails component and created PokemonContext

krzysztof-kryczka committed:
Added navigation component

krzysztof-kryczka committed:
Added userSchema for user

krzysztof-kryczka committed:
installation of necessary lib (@hookform/resolvers)

krzysztof-kryczka committed:
added Registration pages and custom hook useAuth

krzysztof-kryczka committed:
added login page and loginSchema and refactor register page

krzysztof-kryczka committed:
added logout and storage logged user in localstorage

krzysztof-kryczka committed:
added AuthContext

krzysztof-kryczka committed:
added Favorites Page and custom hooks: useFavorites, usePokemonDetails

krzysztof-kryczka committed:
refactor code: Pokemon Details, Pokemon List, shared: Pokemon Card

krzysztof-kryczka committed:
installation of necessary lib (react-icons)

krzysztof-kryczka committed:
big refactor Favorites component division into smaller components, change hooks, added icon heart for favorite and lots more...

krzysztof-kryczka committed:
Added pokemonSchema for CreatePokemon

krzysztof-kryczka committed:
added pages: EditPage and CreatePokemonPage and routing in main.jsx

krzysztof-kryczka committed:
added support for checking if an image has already been used and if so, blocking the ability to create a poke

krzysztof-kryczka committed:
modifications: form now uses react hook form, validation in zod, minor changes in pokemonSchema

krzysztof-kryczka committed:
refactor Create Pokemon

krzysztof-kryczka committed:
merge Pokémon from external API and local db, fix: pokemon.abilities in PokemonInfo

krzysztof-kryczka committed:
fix: updated useFetchPokemons to properly handle local and external Pokémon sources. Refactor useFetchPokemons to correctly calculate offset when combining local and external Pokémon sources

krzysztof-kryczka committed:
feature: editing functionality by name across multiple components. Updated component EditPage to allow editing Pokémon by their unique names. Updated schemaPokémon. Created layout and validation. Dynamically populated form fields with selected Pokémon data.

krzysztof-kryczka committed:
refactor: simplified form handling and reduced code duplication. Combine Edit and Create Pokemon forms into a single reusable component

krzysztof-kryczka committed:
fix: improved visual consistency by using high-quality dream world sprites pokemon.sprites?.other.dream_world.front_default

krzysztof-kryczka committed:
modified: Pokémon data is properly updated or added to db.json based on the id, providing clear user feedback.

krzysztof-kryczka committed:
refactor: Integrate create and edit Pokémon functionality with get sprite into a single save function

krzysztof-kryczka committed:
updated: pagination and context handling for Pokedex and EditPage. Set totalCount in PokemonContext instead of using a default value. Fix: Logic for calculating the number of Pokémon based on totalCount in Pagination.

krzysztof-kryczka committed:
refactor: Component Edit Page + separate component per view

krzysztof-kryczka committed:
feature: RankingPage component + view

krzysztof-kryczka committed:
feature: Implemented sorting functionality for Pokémon list based on user-selected criteria

krzysztof-kryczka committed:
fix: duplicate exclusion by checking local Pokémon IDs (localPokemonIds) and filtering out external Pokémon that already exist in local data.

krzysztof-kryczka committed:
feature: all data fetching logic is now handled centrally in api.js for better maintainability and code reuse, promoting consistency and easier future updates.

krzysztof-kryczka committed:
Modified: api.js and useFavorites to ensure different users can add the same Pokémon to their favorites.
Added error handling in api.js to manage cases where favorite Pokémon data is missing.

krzysztof-kryczka committed:
fix: use useCallback to memoize fetchFavorites function to prevent infinite loop issues. Updated toggleFavorite in useFavorites to remove Pokémonfrom the favorites array if userId array becomes empty.

krzysztof-kryczka committed:
feature: added ArenaPage component and basic styling

krzysztof-kryczka committed:
feature: implemented function to add Pokémon to the arena,
implemented function to remove Pokémon from the arena,
updated: Arena page to show Pokémon on the arena,
logic to correctly show the number of Pokémon on the arena (e.g., 0/2, 1/2, 2/2).
style: hide favorite button on Arena page

krzysztof-kryczka committed:
fix: implement conditional rendering of Remove from Arena button in PokemonCard
added: background image of arena as background
added: removePokemonFromArena function to send a DELETE request to remove Pokémon by id from the arena.

krzysztof-kryczka committed:
updated: save essential Pokémon data  (id, name, weight, height, base_experience, sprite, abilities) to local database before adding to Arena.
Added logic to check if Pokémon already exists in the local database before saved to avoid duplicate entries.

krzysztof-kryczka committed:
feature: Implemented battle logic in ArenaPage to handle Pokémon battles. Updated handleBattle function to determine the winner and loser based on Pokémon experience and weight.

krzysztof-kryczka committed:
added: state management to track the losing Pokémon with loserId. Applied opacity effect to losing Pokémon.

krzysztof-kryczka committed:
feature: added PrivateRoute and PublicRoute components for route protection. Added user-friendly messages in PrivateRoute to inform users trying to access protected routes.

krzysztof-kryczka committed:
removed: redundant authentication checks in RegisterPage, LoginPage, FavoritesPage component.

krzysztof-kryczka committed:
refactor: ensured that Pokémon are fetched in sequence without duplicates on subsequent pages. Optimize while loop in hook to fetch additional Pokémon on per page.
fix: reset additionalPokemonsCount when navigating back to the first page.
fix: use useRef for additionalPokemonsCount to prevent infinite re-renders.
fix: improve pagination in useFetchPokemons, adjusted offset calculations to handle pagination correctly.

krzysztof-kryczka committed:
feature: added theme toggle in navigation component + basic config + themeProvider

krzysztof-kryczka committed:
feature: update handleLeaveArena to remove Pokémon by IDs. Implemented logic to remove each Pokémon from the arena by their IDs after the battle ends.
Iteratively called removePokemonFromArena for each Pokémon in the arena.
Ensured the arena state is cleared and modal is closed after Pokémon removal.
Improved error handling and user feedback in case of failures during Pokémon removal.

krzysztof-kryczka committed:
minor corrections + darkMode in tailwind config (no use prefers-color-scheme)

krzysztof-kryczka committed:
feature: Implemented mobile menu toggle button with FaBars and FaTimes icons. Added custom background classes for light and dark modes.

krzysztof-kryczka committed:
Added dark mode for PokeDex and added conditional rendering for element visibility based on pokemon wins or losses

krzysztof-kryczka committed:
refactor Navigation component.
Implemented 'map()' to generate navigation links dynamically for both authenticated and unauthenticated users.
Consolidated link definitions into arrays 'authenticatedLinks' and 'unauthenticatedLinks'.

krzysztof-kryczka committed:
refactor: Dark Mode + Preliminary Improve mobile responsivenes

krzysztof-kryczka committed:
fix: React Router future flags to resolve warnings

krzysztof-kryczka committed:
updated hook useAuth: login function to store only necessary user information in localStorage (user id, name, email, and isAuthenticated flag).

krzysztof-kryczka committed:
modified FavoritesPage:
added: pagination for better user experience (15 pokémons per page).
used: startIndex and endIndex in useEffect to check for empty current page when remove last pokémon from favorites.
fix: getFavoritesByUserId function to use exact userId matching

krzysztof-kryczka committed:
refactor: useAuth hook and AuthContext for better state management.
isAuthenticated used by leveraging context in PokemonCard and other.

krzysztof-kryczka committed:
created StyledButton component with support for light and dark modes.
Ensured StyledButton component is reusable and consistent with the application design.
Added className prop to StyledButton to allow additional class overrides.
Updated Navigation component to style Link components.
Updated header footer colors in Navigation for light and dark modes.

krzysztof-kryczka committed:
refactor: Navigation component to reduce redundancy using StyledNavLinkButton.
refactor: Arena components.
added: Capitalize first letters of winner and loser names in BattleResultModal
fix: login and register page (StyledButton)

krzysztof-kryczka committed:
refactor: CommonPokemonFormFields to reduce style redundancy.
Created StyledFormField component to handle form labels and input fields.

krzysztof-kryczka committed:
refactor: PokemonCard, PokemonInfo to reduce style redundancy using PokemonStatistics
Improved code readability and maintainability by minimizing repeated styles in more other files

krzysztof-kryczka committed:
refactor: Login and Register User to reduce style redundancy and remove redundant enqueueSnackbar calls in Register and Login Page

krzysztof-kryczka committed:
refactor: EditPage and CreatePokemon to reduce style redundancy

krzysztof-kryczka committed:
refactor: RankingPage reduce style redundancy in PokemonListDisplay using StyledTableCell and StyledTableHeaderCell.
Refactor sort options in RankingPage to use PokemonSort component
Added method sort asc, desc, and label for selected sort method in PokemonSort

krzysztof-kryczka committed:
refactor: reduce style redundancy in more coponents and pages.

krzysztof-kryczka committed:
fix: changed router path from pokemon/:name to pokemon/:id to use ID. Update PokemonDetails to handle ID and add error handling.
Updated links in PokemonList and FavoritesPage to use ID

krzysztof-kryczka committed:
refactor: reduce style redundancy in more coponents and pages.

krzysztof-kryczka committed:
fix: sorting in RankingPage. Added functions handleSortChange and handleOrderChange to reset the page to the first one after changing the sort criteria or sort order. Used totalCount instead of the pages state to calculate the number of pages, ensuring consistency and simplifying the code. Ensured that each Pokémon has a defined wins value by assigning 0 to those who do not have it.

krzysztof-kryczka committed:
fix: major refactor of Pokémon fetching logic to improve performance and consistency.
Data is fetched only when needed and managed properly within context.
Added condition to fetch all Pokémon data if not already available in context.
Updated display local Pokémon in reverse order (newest first).
Updated UI components for better consistency.
These changes enhance the overall user experience by ensuring that the newest Pokémon are always visible first, providing immediate feedback when adding new Pokémon, and maintaining consistency across the application.

krzysztof-kryczka committed:
fix: Forms: Login(user), Register(user), Edit(pokemon), Add(pokemon) : Incorporated useFormContext for better state management across form fields.

krzysztof-kryczka committed:
updated readme.md

krzysztof-kryczka committed:
fix: fixed various bugs

krzysztof-kryczka committed:
fix: Arena Added logging for Pokémon data and improve update logic
fix: Auth Added logging for user data during validation with db.json

RafalKielbasa commented on FrontEnd-service/src/api/api.js:
```diff
@@ -0,0 +1,94 @@
+
+// page: create/edit
+export const fetchUsedSprites = () => localAxios.get('/pokemons')
+//export const savePokemon = pokemonData => localAxios.post('/pokemons', pokemonData)
```
Before submitting for review, unnecessary code should be removed.

RafalKielbasa commented on FrontEnd-service/src/api/api.js:
```diff
@@ -0,0 +1,94 @@
+}
+
+// page: favorites
+// export const getFavoritesByUserId = userId => localAxios.get(`/favorites?userId=${userId}`);
```
Before submitting for review, unnecessary code should be removed.

RafalKielbasa commented on FrontEnd-service/src/assets/pokemons_05.png:
Some of the images are unnecessary.

RafalKielbasa commented on FrontEnd-service/src/components/Loader.jsx:
```diff
@@ -0,0 +1,9 @@
+export const Loader = () => {
+   return (
+      <div className="flex justify-center items-center h-screen">
+         <div className={`border-8 border-gray-200 border-t-8 border-t-blue-500 rounded-full w-28 h-28 animate-spin`} />
```
We don't need curly braces when passing a string as a prop.

RafalKielbasa commented on FrontEnd-service/src/components/Navigation.jsx:
```diff
@@ -0,0 +1,135 @@
+   const { user, isAuthenticated, logout } = useAuth()
+   const { enqueueSnackbar } = useSnackbar()
+   const { theme, toggleTheme } = useContext(ThemeContext)
+   const [menuOpen, setMenuOpen] = useState(false)
```
The state should be placed at the very top of the component.

RafalKielbasa commented on FrontEnd-service/src/components/Navigation.jsx:
I would suggest extracting the conditionally rendered components.

RafalKielbasa commented on FrontEnd-service/src/components/Pagination.jsx:
```diff
@@ -0,0 +1,92 @@
+         {pageType !== 'edit' && pageType !== 'ranking' && pageType !== 'favorites' && (
+            <PaginationButton page={1} currentPage={currentPage} onPageChange={onPageChange} />
+         )}
+         {pageType !== 'edit' &&
```
I suggest conditionally rendering the entire pagination.

krzysztof-kryczka committed:
fix: extract conditionally rendered components from Navigation
- Extracted MenuToggle component to handle menu toggle button.
- Extracted UserStatus component to display user login status.
- Extracted ThemeToggle component to handle theme switching.
- Extracted AuthenticatedLinks component to display links for authenticated users.
- Extracted UnauthenticatedLinks component to display links for unauthenticated users.
- Simplified Navigation component by utilizing the newly extracted components.

fix: conditionally render Pagination component based on pageType
- Added conditional rendering to PaginationButton component based on pageType.
- Simplified Pagination component by grouping conditions and using concise rendering logic.
